### DISCUSSION TITLE
What's the proper way to add telescope extensions?

### DISCUSSION THREAD

**Discussion by kapral18** [LazyVim/LazyVim#1301] [Q&A] [answered]
For example if I want to add aaronhallaert/advanced-git-search.nvim how would  I add it to my lazyvim [config](https://github.com/kapral18/config.nvim)?

**Comment by dpetka2001** [comment]
It has an [example](https://github.com/aaronhallaert/advanced-git-search.nvim#telescope) about how to do it in the docs.

**Reply by kapral18** [reply]
does this

```lua
{
    "aaronhallaert/advanced-git-search.nvim",
    config = function()
        -- optional: setup telescope before loading the extension
        require("telescope").setup{
            -- move this to the place where you call the telescope setup function
            extensions = {
                advanced_git_search = {
                        -- See Config
                    }
            }
        }

        require("telescope").load_extension("advanced_git_search")
    end,
    dependencies = {
        --- See dependencies
    },
}
```

in any way mess up 
https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/editor.lua#L87
https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/editor.lua#L254 

?

**Reply by dpetka2001** [reply]
It shouldn't. Your own spec gets merged with the parent one of Lazyvim.

**Reply by dpetka2001** [reply]
You can take a look at my [telescope](https://github.com/dpetka2001/dotfiles/blob/main/dot_config/nvim/lua/plugins/telescope.lua) configuration. I use 2 extensions and you can see how they are set up, so you can get an idea of how to configure what you want in your own spec.

**Reply by kapral18** [reply, marked as answer]
in the end this worked for me

```lua
return {
  {
    "aaronhallaert/advanced-git-search.nvim",
    config = function()
      require("telescope").setup({
        extensions = {
          advanced_git_search = {
            -- See Config    -- fugitive or diffview
            diff_plugin = "diffview",
          },
        },
      })
      require("telescope").load_extension("advanced_git_search")
    end,
    dependencies = {
      "nvim-telescope/telescope.nvim",
      -- to show diff splits and open commits in browser
      "tpope/vim-fugitive",
      -- to open commits in browser with fugitive
      "tpope/vim-rhubarb",
      -- optional: to replace the diff from fugitive with diffview.nvim
      -- (fugitive is still needed to open in browser)
      "sindrets/diffview.nvim", --- See dependencies
    },
  },
}

```

**Reply by dpetka2001** [reply]
Do you use `vim-fugitive` and `vim-rhubarb`? If not, no need to install them. Glad you worked it out :smile:

**Reply by pidgeon777** [reply]
I think that by doing as suggested, you would overwrite the other `Telescope` custom settings.

I adjusted the configuration, and this seems to work:

```lua
return {
  {
    "aaronhallaert/advanced-git-search.nvim",
    cmd = {
      "AdvancedGitSearch",
      -- WARN: If no further arguments are given to the 'Telescope advanced_git_search' command,
      -- an 'attempt to call a nil value' error will occur!
      -- https://github.com/aaronhallaert/advanced-git-search.nvim/issues/55
      "Telescope advanced_git_search",
    },
    keys = {
      { "<leader>ga", "<cmd>AdvancedGitSearch<cr>", desc = "Advanced git search" },
    },
    dependencies = {
      {
        "telescope.nvim",
        opts = function(_, opts)
          -- Note: taken from 'advanced_git_search' README.md
          local def_opts = {
            -- fugitive or diffview
            diff_plugin = "fugitive",
            -- customize git in previewer
            -- e.g. flags such as { "--no-pager" }, or { "-c", "delta.side-by-side=false" }
            git_flags = {},
            -- customize git diff in previewer
            -- e.g. flags such as { "--raw" }
            git_diff_flags = {},
            -- Show builtin git pickers when executing "show_custom_functions" or :AdvancedGitSearch
            show_builtin_git_pickers = false,
            entry_default_author_or_date = "author", -- one of "author" or "date"

            -- Telescope layout setup
            telescope_theme = {
              function_name_1 = {
                -- Theme options
              },
              function_name_2 = "dropdown",
              -- e.g. realistic example
              show_custom_functions = {
                layout_config = { width = 0.4, height = 0.4 },
              },
            },
          }

          local my_opts = {
            advanced_git_search = {
              -- fugitive or diffview
              diff_plugin = "diffview",
            },
          }

          opts.extensions = vim.tbl_deep_extend("force", opts.extensions or {}, my_opts)

          -- TODO: Test it
          -- Note: to add 'search_log_content' to Telescope builtin pickers
          -- https://github.com/aaronhallaert/advanced-git-search.nvim/issues/55#issuecomment-1708128547
          -- require("telescope.builtin").search_log_content = require("advanced_git_search.telescope.pickers").search_log_content
        end,
      },
      -- to show diff splits and open commits in browser
      "tpope/vim-fugitive",
      -- to open commits in browser with fugitive
      "tpope/vim-rhubarb",
      -- optional: to replace the diff from fugitive with diffview.nvim
      -- (fugitive is still needed to open in browser)
      "sindrets/diffview.nvim",
    },
    config = function()
      -- WARN: needed otherwise 'AdvancedGitSearch' command will not be available
      -- Note: if it is commented out, 'Telecope advanced_git_search' will work anyway
      require("telescope").load_extension("advanced_git_search")
    end,
    enabled = true,
  },
}
```

**Reply by shanesveller** [reply]
IIUC the `config` style in use above may eagerly load `telescope` before you intend to - here's how some of the LazyVim-provided extensions hook in:

https://github.com/LazyVim/LazyVim/blob/5cabc2ff5b52405c0d41f1a06329567e84d15d3b/lua/lazyvim/plugins/editor.lua#L131-L135

**Comment by pidgeon777** [comment]
> IIUC the `config` style in use above may eagerly load `telescope` before you intend to - here's how some of the LazyVim-provided extensions hook in:
> 
> https://github.com/LazyVim/LazyVim/blob/5cabc2ff5b52405c0d41f1a06329567e84d15d3b/lua/lazyvim/plugins/editor.lua#L131-L135

Yes, but also:

https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-customizing-plugin-specs

From the `LazyVim` documentation:

> - cmd: the list of commands will be extended with your custom commands
> - event: the list of events will be extended with your custom events
> - ft: the list of filetypes will be extended with your custom filetypes
> - keys: the list of keymaps will be extended with your custom keymaps
> - opts: your custom opts will be merged with the default opts
> - dependencies: the list of dependencies will be extended with your custom dependencies
>
> **any other property will override the defaults**

So, a new `config` property definition will override the previously defined `Telescope` `config` properties.

**Reply by shanesveller** [reply]
I was imagining making this suggested version the `config` block on the `aaronhallaert/advanced-git-search.nvim` plugin spec rather than clobbering Telescope's - which looks like is already how both of the above snippets up-thread seem to have declared it?

**Reply by pidgeon777** [reply]
Yes, in this case we're defining a `config` section for the `advanced-git-search.nvim` plugin, so this won't override the `Telescope` one.

Instead, the `Telescope` `opts` key will be merged with the other ones, if any.
